File: LeetCode/3301-3400/Problem3333_FindtheOriginalTypedStringII/FindtheOriginalTypedStringII.py

class Solution:
    def possibleStringCount(self, word: str, k: int) -> int:
        MOD = 10**9 + 7

        from collections import deque

        n = len(word)
        groups = []
        i = 0
        while i < n:
            j = i
            while j < n and word[j] == word[i]:
                j += 1
            groups.append(j - i)
            i = j

        max_len = n
        dp = [0] * (max_len + 1)
        dp[0] = 1

        for group_len in groups:
            new_dp = [0] * (max_len + 1)
            prefix = [0] * (max_len + 2)

            for i in range(max_len + 1):
                prefix[i + 1] = (prefix[i] + dp[i]) % MOD

            for i in range(max_len + 1):
                if i - group_len >= 0:
                    new_dp[i] = (prefix[i] - prefix[i - group_len]) % MOD
                else:
                    new_dp[i] = prefix[i]
            dp = new_dp

        return sum(dp[k:]) % MOD


def main():
    sol = Solution()

    # Example 1
    word1 = "aabbccdd"
    k1 = 7
    print(sol.possibleStringCount(word1, k1))  # Output: 5

    # Example 2
    word2 = "aabbccdd"
    k2 = 8
    print(sol.possibleStringCount(word2, k2))  # Output: 1

    # Example 3
    word3 = "aaabbb"
    k3 = 3
    print(sol.possibleStringCount(word3, k3))  # Output: 8

if __name__ == "__main__":
    main()

File: LeetCode/3301-3400/Problem3333_FindtheOriginalTypedStringII/FindtheOriginalTypedStringII.rs
use std::collections::VecDeque;

pub struct Solution;

impl Solution {
    pub fn possible_string_count(word: &str, k: usize) -> i64 {
        const MOD: i64 = 1_000_000_007;
        let chars: Vec<char> = word.chars().collect();
        let n = chars.len();

        // Group consecutive characters
        let mut groups = Vec::new();
        let mut i = 0;
        while i < n {
            let mut j = i;
            while j < n && chars[j] == chars[i] {
                j += 1;
            }
            groups.push(j - i);
            i = j;
        }

        let total_len = n;
        if k > total_len {
            return 0;
        }

        let mut dp = vec![0i64; total_len + 1];
        dp[0] = 1;

        for &group_len in &groups {
            let mut new_dp = vec![0i64; total_len + 1];
            let mut window_sum: i64 = 0;
            let mut window: VecDeque<i64> = VecDeque::new();

            for i in 1..=total_len {
                let add_val = dp[i - 1];
                window.push_back(add_val);
                window_sum = (window_sum + add_val) % MOD;

                if i > group_len {
                    let removed = window.pop_front().unwrap();
                    window_sum = (window_sum - removed + MOD) % MOD;
                }

                new_dp[i] = window_sum;
            }

            dp = new_dp;
        }

        dp[k..=total_len]
            .iter()
            .fold(0, |acc, &val| (acc + val) % MOD)
    }
}

fn main() {
    let param_1 = String::from("aaabbb");
    let param_2 = 3;

    // Correct borrow and cast
    let ret = Solution::possible_string_count(&param_1, param_2 as usize);
    println!("Output: {}", ret); // Expected: 8

    let param_3 = "a".repeat(962) + &"b".repeat(962);
    let param_4 = 1924;
    let ret2 = Solution::possible_string_count(&param_3, param_4 as usize);
    println!("Output: {}", ret2); // Expected: 1
}
